pkg/cgroup/cgroup.go

```diff
@@ -0,0 +1,292 @@
+}
+
+// general
+
+const procCgroups = "/proc/cgroups"
+const procFilesystems = "/proc/filesystems"
+
+const CgroupV1FsType = "cgroup"
+const CgroupV2FsType = "cgroup2"
+const CgroupV1Controller = "cpuset"
+
+const DefaultCgroupV1Path = "/sys/fs/cgroup"         // cgroupv1 only OR cgroupv1 + cgroupv2
+const UnifiedCgroupV2Path = "/sys/fs/cgroup/unified" // cgroupv1 + cgroupv2
+const DefaultCgroupV2Path = "/sys/fs/cgroup"         // cgroupv2 only
+
+const DefaultCgroupV1ReleaseAgent = DefaultCgroupV1Path + CgroupV1Controller + "/release_agent"
+const UnifiedCgroupV2Controllers = UnifiedCgroupV2Path + "/cgroup.controllers"
+const DefaultCgroupV2Controllers = DefaultCgroupV2Path + "/cgroup.controllers"
+
+const TmpCgroup = "/tmp"
+const TmpCgroupV1Path = "tracee-cgroupv1"
+const TmpCgroupV2Path = "tracee-cgroupv2"
+
+const CgroupV2Controllers = TmpCgroupV2Path + "/cgroup.controllers"
+
+// GetCgroupPath walks the cgroup fs and provides the cgroup directory path of
+// given cgroupId and subPath (related to cgroup fs root dir). If subPath is
+// empty, then all directories from cgroup fs will be searched for the given
+// cgroupId.
+func GetCgroupPath(rootDir string, cgroupId uint64, subPath string) (string, error) {
+	entries, err := os.ReadDir(rootDir)
+	if err != nil {
+		return "", err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		entryPath := filepath.Join(rootDir, entry.Name())
+		if strings.HasSuffix(entryPath, subPath) {
+			// Lower 32 bits of the cgroup id == inode number of matching cgroupfs entry
+			var stat syscall.Stat_t
+			if err := syscall.Stat(entryPath, &stat); err == nil {
+				// Check if this cgroup path belongs to cgroupId
+				if (stat.Ino & 0xFFFFFFFF) == (cgroupId & 0xFFFFFFFF) {
+					return entryPath, nil
+				}
+			}
+		}
+
+		// No match at this dir level: continue recursively
+		path, err := GetCgroupPath(entryPath, cgroupId, subPath)
+		if err == nil {
+			return path, nil
+		}
+	}
+
+	return "", fs.ErrNotExist
+}
+
+// Returns a cgroup controller hierarchy value
+func GetCgroupControllerHierarchy(subsys string) (int, error) {
+	var value int
+
+	file, err := os.Open(procCgroups)
+	if err != nil {
+		return -1, fmt.Errorf("could not open %s: %w", procCgroups, err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
+		sline := strings.Fields(scanner.Text())
+		if len(sline) < 2 || sline[0] != subsys {
+			continue
+		}
+		value, err = strconv.Atoi(sline[1])
+		if err != nil || value < 0 {
+			return -1, fmt.Errorf("error parsing %s: %w", procCgroups, err)
+		}
+	}
+
+	return value, nil
+}
+
+// Checks if cgroupv1 is the default cgroup filesystem
+func IsCgroupV1Default() (bool, error) {
+	var err [3]error
+	var cpu, cpuset, memory int
+
+	cpu, err[0] = GetCgroupControllerHierarchy("cpu")
+	cpuset, err[1] = GetCgroupControllerHierarchy("cpuset")
+	memory, err[2] = GetCgroupControllerHierarchy("memory")
+
+	for _, e := range err {
```

Comment:
It can be (although less common) that not all cgroup hierarchies are compiled into the kernel, but here you assume they always are.

Reply:
Yes, but I'm assuming the 3 most important ones are always available. I believe they are in all environments we currently test tracee,

Reply:
@yanivagman, I'm reworking this PR entirely, after refactoring the code in other already merged PRs, but, about:


> I think we should opt to use the previous method for deciding about group v1/v2 since it is the standard way to do this check and also runc uses it (and we need to be consistent with runc): https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/fs2/create.go#L14

> This is important for correctly resolving the container ID from the cgroup path. Also, it will allow not to return an error from this function, and if the file doesn't exist we can simply assume cgroup v1 (as we did with the current implementation)

About hierarchyID, the cgroups man page states:

```
    2. The unique ID of the cgroup hierarchy on which this
       controller is mounted.  If multiple cgroups v1
       controllers are bound to the same hierarchy, then each
       will show the same hierarchy ID in this field.  The
       value in this field will be 0 if:

       a) the controller is not mounted on a cgroups v1
          hierarchy;

       b) the controller is bound to the cgroups v2 single
          unified hierarchy; or

       c) the controller is disabled (see below).
```

So, the previous method (currently in use) always assumes that cgroupv1 is mounted in /sys/fs/cgroup tmpfs and that might not always be the case. We could have unmounted cgroupv1 but, still, being the default, for example.

Kubernetes documentation (https://kubernetes.io/docs/concepts/architecture/cgroups/) is also misleading as they tell user to identify if "/sys/fs/cgroup" is a tmpfs or a cgroup2 filesystem (but in our case the filesystems might not even be mounted).

With that said, I believe the correct way of detecting the default cgroup version being used is through hiearchyID of cpuset from /proc/cgroups.

Reply:
About last comment, reading: https://github.com/systemd/systemd/blob/69c0807432fa4fbfbf507a53872664cd26715559/src/basic/cgroup-util.c#L1986 we see that systemd relies in executing stat on "/sys/fs/cgroup" to check if it is a cgroup2 filesystem (same approach documented by kubernetes). The fact is that we might run tracee inside a container with no group mounted (let's say). Then, the /proc/cgroups is the best alternative to understand which group is the default and to mount one (or both) cgroup versions.